### DISCUSSION TITLE
Considering Traffic Signals in Valhalla Isochrone Analysis

### DISCUSSION THREAD

**Discussion by pipo3944** [valhalla/valhalla#4750] [Q&A]
Hello,

I have a question regarding how traffic signals are considered in Valhalla's isochrone analysis. Specifically, I would like to know if any time loss due to traffic signals is factored in.

Additionally, is it possible to adjust the settings related to this aspect?

Thank you for your assistance.

**Comment by nilsnolde** [comment]
> Additionally, is it possible to adjust the settings related to this aspect?

nope I don't think so, but might be wrong.

yes, traffic lights are factored in, but you'd need to scour the codebase to understand the exact factoring (it's not super straight forward to understand though..)

**Comment by kevinkreiser** [comment]
https://github.com/valhalla/valhalla/blob/4f6ef0595dadf530bc88ae7721af2d010e3de34e/valhalla/sif/osrm_car_duration.h#L58

it looks like currently for automobiles there is only a 2 second time penalty

there is also the notion of `stop_impact` which is increased if a traffic signal exists on the node or edge:

https://github.com/valhalla/valhalla/blob/4f6ef0595dadf530bc88ae7721af2d010e3de34e/src/mjolnir/graphenhancer.cc#L1176

stop impact then becomes a multiplier for a penalty on the transition at the node:

https://github.com/valhalla/valhalla/blob/4f6ef0595dadf530bc88ae7721af2d010e3de34e/src/sif/autocost.cc#L596

**Comment by dnesbitt61** [comment]
Yea, this is somewhat complex. Most of the time penalty is determined by stop impact - which uses road classifications (highway tags) and geometry to infer which transitions are likely to take more time - e.g. a tertiary road crossing a primary road is higher than a primary road crossing a tertiary road. Presence of a traffic light (to me) is less meaningful since one does not know the timing of the light (which directions are favored for instance).

**Comment by pipo3944** [comment]
Thank you all for the detailed information!

I understand now that traffic signals are factored into Valhalla's isochrone analysis, primarily through a 2-second time penalty and the concept of stop impact, which adjusts the penalty based on road classifications and geometry.

However, as you mentioned, predicting the timing of traffic signals seems challenging. Therefore, I will proceed with the default 2-second setting as default.

Thank you once again for your assistance.
